homeassistant/components/zha/__init__.py

```diff
@@ -369,6 +419,123 @@ def zdo_command(self, tsn, command_id, args):
+                self._domain,
+                self._ieeetail,
+                discovery_info['endpoint'].endpoint_id,
+                discovery_info.get('entity_suffix', '')
+            )
+
+
+class ZHASwitchEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0040):
+            self._hass.bus.fire(
+                OFF_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id in (0x0001, 0x0041, 0x0042):
+            self._hass.bus.fire(
+                ON_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id == 0x0002:
+            self._hass.bus.fire(
+                TOGGLE_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+
+
+class ZHALevelEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+        self._level = 0
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0004):  # move_to_level, -with_on_off
+            self.set_level(args[0])
+        elif command_id in (0x0001, 0x0005):  # move, -with_on_off
+            # We should dim slowly -- for now, just step once
+            rate = args[1]
+            if args[0] == 0xff:
+                rate = 10  # Should read default move rate
+            self.move_level(-rate if args[0] else rate)
+        elif command_id == 0x0002:  # step
+            # Step (technically shouldn't change on/off)
+            self.move_level(-args[1] if args[0] else args[1])
+
+    def attribute_update(self, attrid, value):
+        """Handle attribute updates on this cluster."""
+        if attrid == 0:
+            self._level = value
+
+    def move_level(self, change):
+        """Increment the level."""
+        self.set_level(min(255, max(0, self._level + change)))
+
+    def set_level(self, level):
+        """Set the level."""
+        if level == 0 and self._level > 0:
+            self._hass.bus.fire(
+                OFF_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif level > 0 and self._level == 0:
+            self._hass.bus.fire(
+                ON_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        self._level = level
+        self._hass.bus.fire(
+            LEVEL_CHANGE_EVENT_KEY,
```

Comment:
Wouldn't something that has a level be a sensor ?

Reply:
I think that's why this was originally added as a binary sensor. the binary sensor was trying to track on vs. off and level was an attribute.

Reply:
I'm not sure how to feel about "level". On one hand I'd expect event to be stateless, so no actual level value, just "level_up" or "level_down" events, but on the other hand it is nice to have everything wired together out of the box.